### ISSUE TITLE
Unable to use a XML file when dependencies are resolved from workspace projects

### ISSUE THREAD

**Issue by tduchateau** [thymeleaf/thymeleaf-extras-eclipse-plugin#15]
I think I already ran into this issue when I contributed to the plugin some weeks ago.

When the m2e plugin is configured to resolve to dependencies from workspace projects (Right-click on a project > Properties > Maven > "Resolve dependencies ...", the default conf I think), the XML file of the custom dialect is not read, so the content assist is not updated with the custom processors.

When the checkbox in unchecked, the XML file is well read (because the scanning must work when reading from JARs).

It would be great for dialect developers to be able to see changes in the XML file on-the-fly in the content assist.

**Comment by ultraq** [comment]
I take it the custom dialect XML file is in that other Maven project, and that when you uncheck 'Resolve dependencies from workspace projects' there's a JAR within the project that contains the custom dialect XML file as well, but it's from another version of that workspace project?

This might be an m2e thing, preventing the XML file from being resolved via that dialect scanning mechanism that we were using.  If so, I'll need to find out what I need to do to read from an m2e resolved project.

**Comment by tduchateau** [comment]
To be more precise, here is my use case :

```
datatables-thymeleaf-0.8.9
|__ src
    |__ main
        |__ java
        |__ resources
            |__ Datatables-Dialect.xml <= my custom dialect
|__ pom.xml

sample-thymeleaf
|__ src
    |__ main
        |__ java
        |__ resources
        |__ webapp
            |__ WEB-INF
                |__ views
                    |__ myView.html <= the view where I'm trying to use the content assist
|__ pom.xml
    |__ a dependency to datatables-thymeleaf-0.8.9 is added in this pom
```

From what I could recall, the scanning feature get all opened projects in the current Eclipse workspace. So I'm not sure if there's an actual distinction between m2e and non-m2e projects. :-/

**Comment by ultraq** [comment]
I remember altering the dialect scanning code to only look at dependencies of the project instead of _all_ projects in an Eclipse workspace (scanning every project on my machine took a long time, maybe 30 seconds, because of all the open Eclipse plugin projects I had imported at the time).  I've sped it up since then by making it multi-threaded, so maybe scanning all projects won't be so bad now?

There must be some middle ground though, a way to resolve m2e projects as well as normal project dependencies, so that scanning doesn't take so long.

**Comment by tduchateau** [comment]
Very weird :
- m2e projects not resolved from workspace are scanned (JAR)
- m2e projects resolved from workspace are not scanned
- "normal" projects are not scanned

I'm using Eclipse Juno (3.8) / WTP (including Web Developer Tools 3.4.1.xxx) / JDK 1.6.0_34

**Comment by ultraq** [comment]
Note to self: also check the 'weirdness' noted in #16.

**Comment by ultraq** [comment]
Added support for obtaining dialect files from project dependencies, either through managed containers (eg: Maven, Gradle), or as projects added to the build path in Eclipse.  Also added very basic support for refreshing of dialect suggestions when the dialect file those suggestions were created from changes.

The current cache and refresh mechanism isn't able to handle more complicated workspace changes (eg: project deletions that had dialect files, getting dialect files when switching between container managed / Eclipse managed project dependencies, creating new dialect files after the initial load), and will need a bit of rework.  I've opened #21 as a refactoring job for the cache/refresh to better handle a wider range of workspace change use cases.
